middleware/humaz/check.go

```diff
@@ -0,0 +1,235 @@
+func WithObjectIDMapper(mapper StringMapper) CheckOption {
+	return func(o *CheckOptions) {
+		o.obj.idMapper = mapper
+	}
+}
+
+// WithObjectIDFromVar takes the name of a variable in the request path that is used as the object id to check.
+func WithObjectIDFromVar(name string) CheckOption {
+	return func(o *CheckOptions) {
+		o.obj.idMapper = func(ctx huma.Context) string {
+			return ctx.Param(name)
+		}
+	}
+}
+
+// WithObjectMapper takes a function that is used to determine the object type and id to check from the incoming request.
+func WithObjectMapper(mapper ObjectMapper) CheckOption {
+	return func(o *CheckOptions) {
+		o.obj.mapper = mapper
+	}
+}
+
+// WithPolicyPath sets the path of the policy module to use for the check call.
+func WithPolicyPath(path string) CheckOption {
+	return func(o *CheckOptions) {
+		o.policy.path = path
+	}
+}
+
+// CheckOptions is used to configure the check middleware.
+type CheckOptions struct {
+	obj struct {
+		id       string
+		objType  string
+		idMapper StringMapper
+		mapper   ObjectMapper
+	}
+	rel struct {
+		name   string
+		mapper StringMapper
+	}
+	subj struct {
+		subjType string
+		mapper   IdentityMapper
+	}
+	policy struct {
+		path   string
+		mapper StringMapper
+	}
+}
+
+func (o *CheckOptions) object(ctx huma.Context) (string, string) {
+	objType := o.obj.objType
+	objID := o.obj.id
+
+	switch {
+	case o.obj.mapper != nil:
+		objType, objID = o.obj.mapper(ctx)
+	case o.obj.idMapper != nil:
+		objID = o.obj.idMapper(ctx)
+	}
+
+	return objType, objID
+}
+
+func (o *CheckOptions) relation(g huma.Context) string {
+	relation := o.rel.name
+	if o.rel.mapper != nil {
+		relation = o.rel.mapper(g)
+	}
+
+	return relation
+}
+
+func (o *CheckOptions) subjectType() string {
+	if o.subj.subjType != "" {
+		return o.subj.subjType
+	}
+
+	return "user"
+}
+
+type Check struct {
+	mw   *Middleware
+	opts *CheckOptions
+}
+
+func newCheck(mw *Middleware, options ...CheckOption) *Check {
+	opts := &CheckOptions{}
+	for _, o := range options {
+		o(opts)
+	}
+
+	return &Check{mw: mw, opts: opts}
+}
+
+// Handler returns a middleware handler that checks incoming requests.
+func (c *Check) Handler(ctx huma.Context, next func(huma.Context)) {
+	policyContext := c.policyContext(ctx)
+	identityContext := c.identityContext(ctx)
+
+	resourceContext, err := c.resourceContext(ctx)
+	if err != nil {
+		ctx.SetStatus(http.StatusInternalServerError)
+		return
+	}
+
+	allowed, err := c.mw.is(ctx.Context(), identityContext, policyContext, resourceContext)
+	if err != nil {
+		ctx.SetStatus(http.StatusInternalServerError)
+		return
+	}
+
+	if !allowed {
+		ctx.SetStatus(http.StatusForbidden)
+		return
+	}
+
+	next(ctx)
+}
+
+// Allowed returns a boolean indicating whether the request is allowed or not.
+// It returns false if the request is not allowed or an error in the check happens.
+func (c *Check) Allowed(ctx huma.Context) (bool, error) {
```

Comment:
Does this function be exported? It looks like it should be called internally by `Handler`. Something like:
```go
// Handler returns a middleware handler that checks incoming requests.
func (c *Check) Handler(ctx huma.Context, next func(huma.Context)) {
    allowed, err := c.allowed(ctx)

    switch {
    case err != nil:
        ctx.SetStatus(http.StatusInternalServerError)
    case !allowed:
        ctx.SetStatus(http.StatusForbidden)
    default:
        next(ctx)
    }
}
```

Reply:
I would like to keep it exported because we have a specific requirement to only know if it's allowed or not so that we can act based on the decision that was returned and enforce it ourself. But i will change the `Handler` function so that it calls the `Allowed` function.